src/Microsoft.AspNet.Routing/RouteCollection.cs

```diff
@@ -197,7 +198,31 @@ public virtual string GetVirtualPath(VirtualPathContext context)
             }
         }
 
-        private void EnsureLogger(HttpContext context)
+        private string NormalizeVirtualPath(String url)
+        {
+            if (string.IsNullOrEmpty(url)) return url;
+            if (_options.LowercaseUrls)
+            {
+                var lowercaseUrl = url;
+                var queryString = "";
+
+                int index = url.IndexOfAny(new char[] { '?', '#' });
+                if (index >= 0)
```

Comment:
Personally I'd find it more readable by doing all the work in one `if()`, basically like this:

```c#
int indexOfSeparator = url.IndexOfAny(...);
if (indexOfSeparator == -1)
{
    // no separator, just lowercase the whole thing
    ...
}
else
{
    // there's a separator, so do some fancy splitting and assembling
    ...
}
```